govcd/nsxv_firewall.go

```diff
@@ -0,0 +1,235 @@
+/*
+ * Copyright 2019 VMware, Inc.  All rights reserved.  Licensed under the Apache v2 License.
+ */
+
+package govcd
+
+import (
+	"encoding/xml"
+	"fmt"
+	"net/http"
+
+	"github.com/vmware/go-vcloud-director/v2/types/v56"
+	"github.com/vmware/go-vcloud-director/v2/util"
+)
+
+// requestEdgeFirewallRules nests EdgeFirewallRule as a convenience for unmarshalling POST requests
+type requestEdgeFirewallRules struct {
+	XMLName           xml.Name                  `xml:"firewallRules"`
+	EdgeFirewallRules []*types.EdgeFirewallRule `xml:"firewallRule"`
+}
+
+// responseEdgeFirewallRules is used to unwrap response when retrieving
+type responseEdgeFirewallRules struct {
+	XMLName           xml.Name                 `xml:"firewall"`
+	Version           string                   `xml:"version"`
+	EdgeFirewallRules requestEdgeFirewallRules `xml:"firewallRules"`
+}
+
+// CreateNsxvFirewall creates firewall rule using proxied NSX-V API. It is a synchronuous operation.
+// It returns an object with all fields populated (including ID)
+// If aboveRuleId is not empty, it will send a query parameter aboveRuleId= which instructs NSX to
+// place this rule above the specified rule ID
+func (egw *EdgeGateway) CreateNsxvFirewall(firewallRuleConfig *types.EdgeFirewallRule, aboveRuleId string) (*types.EdgeFirewallRule, error) {
```

Comment:
Similarly to the Terraform PR, I believe all the related functions in govcd should contain `...FirewallRule` at the end. We are dealing with firewall rules here, as per comment above, hence we should name the function as such.

Note: `GetFwGeneralParams()`, `UpdateFwGeneralParams()` are probably exceptions, because they do deal with the whole FW config at once.

CC: @vbauzysvmware @dataclouder

Reply:
I made same note in my review, also recommend to change from `Params` to `Config`

Reply:
This is changed. `GetFwGeneralParams()` and `UpdateFwGeneralParams()` have become `GetFwGeneralConfig()` and `UpdateFwGeneralConfig()` respectively.